### PR TITLE
Add kvcache support to debug_analysis.py and engine.py

### PR DESCRIPTION
**Pull request by alexm-nm**
Adds the ability to run kvcache via the debug_analysis.py.

### CODE REVIEW

File: src/deepsparse/debug_analysis.py [hunk 1]
```diff
@@ -295,11 +314,36 @@ def main():
...
         with override_onnx_input_shapes(model_path, input_shapes) as tmp_path:
-            input_list = generate_random_inputs(tmp_path, args.batch_size)
+            input_list = generate_random_inputs(tmp_path, batch_size)
     else:
-        input_list = generate_random_inputs(model_path, args.batch_size)
+        input_list = generate_random_inputs(model_path, batch_size)
+
+    kv_cache_params = None
+    if args.use_kvcache:
+        if args.kv_cache_prev_num_tokens is None:
```
Comment (SageMoore) [inline src/deepsparse/debug_analysis.py:330]: Defaulting both of these to zero is fine. Especially for benchmarking.

**Review by SageMoore — COMMENTED** [review]
This diff looks reasonable to me, Alex. I think once we resolve the formatting issues we should be good to merge.

File: src/deepsparse/debug_analysis.py (hunk 1 above)
Comment (alexm-nm) [reply src/deepsparse/debug_analysis.py:330]: Good idea to reduce command size. Added default 0s.

**Comment by alexm-nm** [comment]
Formatting issues are resolved (thanks to Michael)

File: src/deepsparse/debug_analysis.py [hunk 2]
```diff
@@ -140,6 +140,29 @@ def parse_args():
...
+        "--kv-cache-prev-num-tokens",
+        help="KVCache: The amount of previous tokens that will be read"
+        " from the external KV cache on the first inference",
+        type=int,
+        default=None,
+    )
+    parser.add_argument(
+        "--kv-cache-num-frozen-tokens",
+        help="KVCache: The amount of first tokens that we want to keep"
+        " permanently in the KV cache",
```
Comment (bnellnm) [inline src/deepsparse/debug_analysis.py:162]: can you clarify this comment?  does it mean the first N tokens will be frozen?

Comment (alexm-nm) [reply src/deepsparse/debug_analysis.py:162]: yeah, exactly that. For some reason we need this for OPT models

File: src/deepsparse/engine.py [hunk 3]
```diff
@@ -205,6 +204,43 @@ def __repr__(self) -> str:
...
+    assert len(outputs) > 0
+
+    # Create a boolean list of every output of the
+    # model [logits, key0, value0, key1, value1, ..., keyN, valueN]
+    cached_outputs = [True for i in range(len(outputs))]
+
+    # Assume first input is logits and logits ought not to be cached
+    cached_outputs[0] = False
+
+    return cached_outputs
```
Comment (bnellnm) [inline src/deepsparse/engine.py:223]: i think this should get pushed down into utils/onnx.py

File: src/deepsparse/engine.py [hunk 4]
```diff
@@ -804,10 +860,14 @@ def analyze(
             are setup correctly for the DeepSparse Engine
         :return: the analysis structure containing the performance details of each layer
         """
-        if val_inp:
-            self._validate_inputs(inp)
 
-        [out, bench_info] = self._eng_net.benchmark_execute(inp)
+        if self._kv_cache_params:
+            [_, bench_info] = self._eng_net.benchmark_execute(inp, self._kv_cache)
```
Comment (bnellnm) [inline src/deepsparse/engine.py:865]: shouldn't we validate inputs here too?

**Review by bnellnm — COMMENTED** [review]
Overall looks good.  Just a few minor comments

**Review by bnellnm — CHANGES_REQUESTED** [review]
I meant to hit "request changes" on my last round of comments.  I couldn't figure out how to go back and change it.

File: src/deepsparse/engine.py (hunk 4 above)
Comment (alexm-nm) [reply src/deepsparse/engine.py:865]: Yeah you right. I forgot to add awareness of kvcache to the validate_inputs(..) method. Added it now and merged the code with a None

File: src/deepsparse/engine.py (hunk 3 above)
Comment (alexm-nm) [reply src/deepsparse/engine.py:223]: moved

File: src/deepsparse/engine.py [hunk 5]
```diff
@@ -22,6 +22,7 @@
 from typing import Dict, Iterable, List, Optional, Tuple, Union
 
 import numpy
+import onnx
```
Comment (bnellnm) [inline src/deepsparse/engine.py:25]: nit: don't need this import anymore

File: src/deepsparse/utils/onnx.py [hunk 6]
```diff
@@ -472,3 +473,22 @@ def truncate_onnx_embedding_model(
...
     return output_filepath, tmp_file
+
+
+def default_cached_outputs(model_path):
+    """
+    :param model_path: Path to a model
+    :return A list of bools that indicates caching of all outputs except the first one.
+    """
+
+    outputs = list(onnx.load(model_path).graph.output)
```
Comment (bnellnm) [inline src/deepsparse/utils/onnx.py:484]: if a model has overridden input shapes is that going to affect the outputs here?

**Review by bnellnm — DISMISSED** [review]

File: src/deepsparse/utils/onnx.py (hunk 6 above)
Comment (mgoin) [reply src/deepsparse/utils/onnx.py:484]: nope, it's good we can just look at the outputs

File: src/deepsparse/utils/onnx.py [hunk 7]
```diff
@@ -472,3 +473,22 @@ def truncate_onnx_embedding_model(
     )
 
     return output_filepath, tmp_file
+
+
+def default_cached_outputs(model_path):
```
Comment (mgoin) [inline src/deepsparse/utils/onnx.py:478]:
```suggestion
def default_cached_outputs(model_path: str) -> List[bool]:
```

**Review by bnellnm — APPROVED** [review]

**Review by mgoin — APPROVED** [review]

**Pull request merged**